File: dart/transaction_system/dart_application_1/bin/service_locator.dart
// lib/utils/service_locator.dart

import 'database.dart';
import 'inventory_repository.dart';
import 'transaction_repository.dart';
import 'audit_log_repository.dart';
import 'audit_logger.dart';
import 'inventory_service.dart';
import 'transaction_service.dart';

/// Simple service locator. Replace with get_it package for larger apps.
class ServiceLocator {
  static ServiceLocator? _instance;
  static ServiceLocator get instance => _instance ??= ServiceLocator._init();

  late final InMemoryDatabase db;
  late final InventoryRepository inventoryRepo;
  late final TransactionRepository transactionRepo;
  late final AuditLogRepository auditLogRepo;
  late final AuditLogger auditLogger;
  late final InventoryService inventoryService;
  late final TransactionService transactionService;

  ServiceLocator._init() {
    db = InMemoryDatabase();
    inventoryRepo = InventoryRepository(db);
    transactionRepo = TransactionRepository(db);
    auditLogRepo = AuditLogRepository(db);
    auditLogger = AuditLogger(auditLogRepo);
    inventoryService = InventoryService(
      repo: inventoryRepo,
      auditLogger: auditLogger,
    );
    transactionService = TransactionService(
      inventoryRepo: inventoryRepo,
      transactionRepo: transactionRepo,
      auditLogger: auditLogger,
    );
  }
}

File: python system/inventory_system/utils/service_locator.py
from repositories.database import InMemoryDatabase
from repositories.inventory_repository import InventoryRepository
from repositories.transaction_repository import TransactionRepository
from repositories.audit_log_repository import AuditLogRepository
from services.audit_logger import AuditLogger
from services.inventory_service import InventoryService
from services.transaction_service import TransactionService


class ServiceLocator:
    _instance = None

    def __new__(cls):
        if cls._instance is None:
            cls._instance = super().__new__(cls)
            cls._instance._initialized = False
        return cls._instance

    def __init__(self):
        if self._initialized:
            return
        self._initialized = True

        # Initialize database and repositories
        self.db = InMemoryDatabase()
        self.inventory_repo = InventoryRepository(self.db)
        self.transaction_repo = TransactionRepository(self.db)
        self.audit_log_repo = AuditLogRepository(self.db)

        # Initialize services
        self.audit_logger = AuditLogger(self.audit_log_repo)
        self.inventory_service = InventoryService(
            repo=self.inventory_repo,
            audit_logger=self.audit_logger,
        )
        self.transaction_service = TransactionService(
            inventory_repo=self.inventory_repo,
            transaction_repo=self.transaction_repo,
            audit_logger=self.audit_logger,
        )